dra-evolution/pkg/api/claim_types.go

```diff
@@ -361,9 +401,12 @@ type RequestAllocationResult struct {
 	// device to be allocated.
 	RequestName string `json:"requestName"`
 
-	// DeviceName references one device instance via its name in the resource
-	// pool. Driver name and node name have to be determined from the
-	// context.
+	// This node name together with the driver name and
+	// the device name field identify which device was allocated.
+	NodeName string `json:"nodeName"`
```

Comment:
Let's generalize this a bit more so that it also works for network-attached devices, whether they get allocated through a control plane controller (this would be possible in 1.31) or with some future extension of structured parameters.

```suggestion
	// This name together with the driver name and
	// the device name field identify which device was allocated.
	//
	// For a driver with node-local devices, it is the name of the node
	// and the overall ID is `<driver name>/<segment name = node name>/<device name>`.
	//
	// For a driver where device names are unique in a certain subset of
	// the cluster (region, rack, etc.), the name identifies that segment.
	//
	// For a driver where device names are unique in the cluster, this
	// name may be empty.
	// 
	// +optional
	SegmentName string `json:"segmentName,omitempty"`
```

Reply:
I'm torn between allowing just one intermediate segment and many. It could make sense to allow `<driver name>/<region>/<data center>/<device name>` as the string ID for a device.

We want to be sure that there is a driver name and device name, but what the driver puts into the middle doesn't really matter.

Naming, oh why do you have to be so difficult?!

Using "PoolName" is tempting ("names the pool in which the device name must be unique"), but it clashes with `ResourcePool.Name`.

`DevicePoolName`? Max length 128 characters, DNS sub-domains separated by slashes. Optional.

That goes well together with `DeviceName`.

Once we get to structured parameters for network-attached devices, a `ResourcePool` object for those would have:
- a `NodeSelector *v1.NodeSelector instead of `NodeName string`.
- such a `DevicePoolName` that serves as part of the unique ID

Reply:
We have to allow a node name, so the max length has to be 253.

Reply:
I've pushed this as a third commit.

Reply:
And here's how this would work together with a ResourcePool (but note [my question about that name](https://github.com/kubernetes-sigs/wg-device-management/pull/28/files#r1630964004):

https://github.com/kubernetes-sigs/wg-device-management/blob/78e7edd841422bd887e61061778213ad1e2ea949/dra-evolution/pkg/api/capacity_types.go#L9-L21

https://github.com/kubernetes-sigs/wg-device-management/blob/78e7edd841422bd887e61061778213ad1e2ea949/dra-evolution/pkg/api/capacity_types.go#L50-L102

Reply:
I find it super confusion to have both ResourcePool and DevicePool. I think we can find a better name.

Based on the discussion in the other comment, the purpose of this is to allow the controller to identify the "owner" of the device. In the case of node-local, that is always the node so we didn't need anything but node name. With networked devices, this is controller-specific and effectively an opaque (to K8s) identifier. I can see where you are going with "device pool" in that sense, but I think it's too confusing. So, let's brainstorm some other ideas. What about something of this form: `{"",Device,Resource} + {Manager,Owner,Driver} + {Id,Identifier,Key,Token,OpaqueId}`. So, like, `DriverOpaqueId` would be one possible name based on that. Other ideas?

Reply:
Actually I really like `DriverOpaqueId`. It makes it clear it is 1) driver-specific; 2) opaque to k8s (not a resource name for example); 3) an identifier

Reply:
I agree, the naming of ResourcePool needs further thought. The current name implies "this is one pool" but it's really just a subset of some other, larger pool, the one which contains all devices provided by the same driver name and published with the same "DriverOpaqueID" (or whatever we call it).

I would stay with the ResourceSlice type and define it like this:
```
Devices are identified by the name of the driver, a pool name, and a unique name inside that pool. A driver publishes information about its resources in one or more ResourceSlice objects. Each slice is part of one pool and contains the common attributes of a pool (the pool name, which nodes have access to the devices in the pool) and some of the devices in the pool. The common attributes are provided in each slice to enable correlating them (via the driver name and pool name) and to provide all relevant information without having to look up other objects.

Another common attribute is the total number of devices in the pool. This information enables consumers to determine whether they have seen all slices of a pool.

A slices belonging to the same pool must share the same generation number (TODO: can we use ObjectMeta.Generation for this or do we have to define our own field?). When a consumer sees slices of different generations, it should ignore all old slices. This can happen while the driver is updating the pool.

Pools that are local to a node are special. They get published with a node name, which then also serves as pool name. There are special permission checks that limit access from a node to slices with the same node name. Provisioning a new node of the same type is likely to provide the same set of devices as some old node.
```

With this definition, we can safely implement "give me access to all devices" because scheduler knows when it lacks information. The "pool name" is meaningful and won't get confused with the name of a ResourcePool.

This begs the question whether we should have a separate "pool" type. I don't think we need it, some small amount of duplicated information IMHO is fine.

Reply:
Ahh...yes, this makes sense to me. What if we took this a little further, and we *do* have two types: DevicePool and ResourceSlice. Instead of "Pools that are local to a node are special. They get published with a node name, which then also serves as pool name.", we could have the idea of a "DriverInstance" which could be the node name for node-local devices, but could vary with other types of devices. So, it would be the four values: `(DriverName, DriverInstance, DevicePoolName, DeviceName)` that identifies an individual device. As you say, `ResourceSlice` would be the mechanism by which we publish the *elements* of the DevicePool. We could publish a DevicePool resource as well, and it would include a total device count and/or a slice count, and a generation ID as you say, so that we have this one object that we can use to ensure consistency across all the slices for that pool. We could also normalize common attributes up to the pool, as we have discussed before; I don't think we would need to normalize them up to the *slice*, just to the *pool*.

So - I think this is a good model. What I really like is it weakens the "node" boundary in a way I think will be important in the long run (it's even in our charter...). What I *don't* like, is that it requires a lot of explanation. I think if we omit the DevicePool object itself, it will be harder for people to understand. So I would want to have that in there. I think it also helps with revision control - it lets us know what the current generation is. Otherwise, I imagine some race conditions when we just say it is "the latest one we have seen".

Reply:
I've added these ideas with a different wording to https://github.com/kubernetes/enhancements/pull/4709.

Let's discuss there.